contracts/utils/cryptography/draft-ERC7739Utils.sol

```diff
@@ -0,0 +1,277 @@
+        bytes calldata contentsType,
+        bytes calldata contentsTypeName
+    ) internal pure returns (bytes32) {
+        return
+            keccak256(
+                abi.encodePacked(
+                    "TypedDataSign(",
+                    contentsTypeName,
+                    " contents,bytes1 fields,string name,string version,uint256 chainId,address verifyingContract,bytes32 salt,uint256[] extensions)",
+                    contentsType
+                )
+            );
+    }
+
+    /**
+     * @dev Try to validate the contents type is a valid EIP-712 type.
+     *
+     * A valid `contentsType` is considered invalid if it's empty or if:
+     * - it starts with a '('.
+     * - it contains any of the following bytes: , )\x00
+     *
+     * NOTE: This is a looser take on the ERC very strict restrictions. This part appears to be under discussion, and
+     * therefore the restrictions implemented here may change in a future release.
+     */
+    function tryValidateContentsType(
+        bytes calldata contentsType
+    ) internal pure returns (bool valid, bytes calldata contentsTypeName) {
+        // Loop over the contentsType, looking for the end of the contntsTypeName and validating the input as we go.
+        for (uint256 i = 0; i < contentsType.length; ++i) {
+            bytes1 current = contentsType[i];
+            if (current == bytes1("(")) {
+                // we found the end of the contentsTypeName
+                if (i == 0) {
+                    return (false, _emptyCalldataBytes());
+                } else {
+                    return (true, contentsType[:i]);
+                }
+            } else if (current == 0x00 || current == bytes1(" ") || current == bytes1(",") || current == bytes1(")")) {
+                // we found an invalid character (forbidden)
+                return (false, _emptyCalldataBytes());
+            }
+        }
+        // We exited the loop without finding of the contentsTypeName
+        return (false, _emptyCalldataBytes());
+    }
+
+    /**
+     * @dev Computes the hash of the nested struct for the given contents.
+     *
+     * NOTE: This function does not validate the contents type. See {tryValidateContentsType}.
+     */
+    function typedDataNestedStructHash(
+        bytes calldata contentsType,
+        bytes32 contents,
+        bytes1 fields,
+        string memory name,
+        string memory version,
+        address verifyingContract,
+        bytes32 salt,
+        uint256[] memory extensions
+    ) internal view returns (bytes32) {
+        (, bytes calldata contentsTypeName) = tryValidateContentsType(contentsType);
+        return
+            keccak256(
+                abi.encode(
+                    NESTED_TYPED_DATA_TYPEHASH(contentsType, contentsTypeName),
+                    contents,
+                    fields,
+                    keccak256(bytes(name)),
+                    keccak256(bytes(version)),
+                    block.chainid,
```

Comment:
Is there any use case where this library is used to perform a verification/validate a `typedDataNestedStructHash` for another chain ?